File: src/backend/orcamentos/src/model/orcamento_fluxo_caixa_detalhe_model.py
from src import db
from src.model.fin_natureza_financeira_model import FinNaturezaFinanceiraModel


class OrcamentoFluxoCaixaDetalheModel(db.Model):
    __tablename__ = 'orcamento_fluxo_caixa_detalhe'

    id = db.Column(db.Integer, primary_key=True)
    periodo = db.Column(db.String(10))
    valor_orcado = db.Column(db.Float)
    valor_realizado = db.Column(db.Float)
    taxa_variacao = db.Column(db.Float)
    valor_variacao = db.Column(db.Float)
    id_orcamento_fluxo_caixa = db.Column(db.Integer, db.ForeignKey('orcamento_fluxo_caixa.id'))
    id_fin_natureza_financeira = db.Column(db.Integer, db.ForeignKey('fin_natureza_financeira.id'))

    fin_natureza_financeira_model = db.relationship('FinNaturezaFinanceiraModel', foreign_keys=[id_fin_natureza_financeira])

    def mapping(self, data):
        self.id = data.get('id')
        self.id_orcamento_fluxo_caixa = data.get('idOrcamentoFluxoCaixa')
        self.id_fin_natureza_financeira = data.get('idFinNaturezaFinanceira')
        self.periodo = data.get('periodo')
        self.valor_orcado = data.get('valorOrcado')
        self.valor_realizado = data.get('valorRealizado')
        self.taxa_variacao = data.get('taxaVariacao')
        self.valor_variacao = data.get('valorVariacao')

    def serialize(self):
        return {
            'id': self.id,
            'idOrcamentoFluxoCaixa': self.id_orcamento_fluxo_caixa,
            'idFinNaturezaFinanceira': self.id_fin_natureza_financeira,
            'periodo': self.periodo,
            'valorOrcado': self.valor_orcado,
            'valorRealizado': self.valor_realizado,
            'taxaVariacao': self.taxa_variacao,
            'valorVariacao': self.valor_variacao,
            'finNaturezaFinanceiraModel': self.fin_natureza_financeira_model.serialize() if self.fin_natureza_financeira_model else None,
        }

File: frontend/orcamentos/lib/app/data/model/orcamento_fluxo_caixa_detalhe_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:orcamentos/app/data/model/model_imports.dart';

class OrcamentoFluxoCaixaDetalheModel {
	int? id;
	int? idOrcamentoFluxoCaixa;
	int? idFinNaturezaFinanceira;
	String? periodo;
	double? valorOrcado;
	double? valorRealizado;
	double? taxaVariacao;
	double? valorVariacao;
	FinNaturezaFinanceiraModel? finNaturezaFinanceiraModel;

	OrcamentoFluxoCaixaDetalheModel({
		this.id,
		this.idOrcamentoFluxoCaixa,
		this.idFinNaturezaFinanceira,
		this.periodo,
		this.valorOrcado,
		this.valorRealizado,
		this.taxaVariacao,
		this.valorVariacao,
		this.finNaturezaFinanceiraModel,
	});

	static List<String> dbColumns = <String>[
		'id',
		'periodo',
		'valor_orcado',
		'valor_realizado',
		'taxa_variacao',
		'valor_variacao',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Periodo',
		'Valor Orcado',
		'Valor Realizado',
		'Taxa Variacao',
		'Valor Variacao',
	];

	OrcamentoFluxoCaixaDetalheModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idOrcamentoFluxoCaixa = jsonData['idOrcamentoFluxoCaixa'];
		idFinNaturezaFinanceira = jsonData['idFinNaturezaFinanceira'];
		periodo = jsonData['periodo'];
		valorOrcado = jsonData['valorOrcado']?.toDouble();
		valorRealizado = jsonData['valorRealizado']?.toDouble();
		taxaVariacao = jsonData['taxaVariacao']?.toDouble();
		valorVariacao = jsonData['valorVariacao']?.toDouble();
		finNaturezaFinanceiraModel = jsonData['finNaturezaFinanceiraModel'] == null ? FinNaturezaFinanceiraModel() : FinNaturezaFinanceiraModel.fromJson(jsonData['finNaturezaFinanceiraModel']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idOrcamentoFluxoCaixa'] = idOrcamentoFluxoCaixa != 0 ? idOrcamentoFluxoCaixa : null;
		jsonData['idFinNaturezaFinanceira'] = idFinNaturezaFinanceira != 0 ? idFinNaturezaFinanceira : null;
		jsonData['periodo'] = periodo;
		jsonData['valorOrcado'] = valorOrcado;
		jsonData['valorRealizado'] = valorRealizado;
		jsonData['taxaVariacao'] = taxaVariacao;
		jsonData['valorVariacao'] = valorVariacao;
		jsonData['finNaturezaFinanceiraModel'] = finNaturezaFinanceiraModel?.toJson;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idOrcamentoFluxoCaixa = plutoRow.cells['idOrcamentoFluxoCaixa']?.value;
		idFinNaturezaFinanceira = plutoRow.cells['idFinNaturezaFinanceira']?.value;
		periodo = plutoRow.cells['periodo']?.value;
		valorOrcado = plutoRow.cells['valorOrcado']?.value?.toDouble();
		valorRealizado = plutoRow.cells['valorRealizado']?.value?.toDouble();
		taxaVariacao = plutoRow.cells['taxaVariacao']?.value?.toDouble();
		valorVariacao = plutoRow.cells['valorVariacao']?.value?.toDouble();
		finNaturezaFinanceiraModel = FinNaturezaFinanceiraModel();
		finNaturezaFinanceiraModel?.descricao = plutoRow.cells['finNaturezaFinanceiraModel']?.value;
	}	

	OrcamentoFluxoCaixaDetalheModel clone() {
		return OrcamentoFluxoCaixaDetalheModel(
			id: id,
			idOrcamentoFluxoCaixa: idOrcamentoFluxoCaixa,
			idFinNaturezaFinanceira: idFinNaturezaFinanceira,
			periodo: periodo,
			valorOrcado: valorOrcado,
			valorRealizado: valorRealizado,
			taxaVariacao: taxaVariacao,
			valorVariacao: valorVariacao,
		);			
	}

	
}